### ISSUE TITLE
physicalmemory _constraints useless

### ISSUE THREAD

**Issue by rguenth** [openSUSE/open-build-service#4433]
As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have

`
<constraints>
...
  <overwrite>
   <conditions>
    <package>gcc8</package>
   </conditions>
   <hardware>
     <processors>4</processors>
     <physicalmemory>
      <size unit="G">2</size>
     </physicalmemory>
   </hardware>
  </overwrite>
</constraints>
`

thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
'processors' includes threads and thus matches %{jobs} will evaluate to and thus
the make parallelism that is used.

So please change the beahvior of pyhsicalmemory to scale with the number of threads
in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
this behaivor.

Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
unreasonably (I don't remember if we can specify constraints per architecture).

**Comment by adrianschroeter** [comment]
On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> 
> <constraints>
> ...
>   <overwrite>
>    <conditions>
>     <package>gcc8</package>
>    </conditions>
>    <hardware>
>      <processors>4</processors>
>      <physicalmemory>
>       <size unit="G">2</size>
>      </physicalmemory>
>    </hardware>
>   </overwrite>
> </constraints>
> 
> thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> the make parallelism that is used.
> 
> So please change the beahvior of pyhsicalmemory to scale with the number of threads
> in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> this behaivor.
>
> Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> unreasonably (I don't remember if we can specify constraints per architecture).

It is the task of the spec file to specify a maximum number of parallel jobs, but
this is independend of the available hardware.

Also I don't want to do an incompatible change here.


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by rguenth** [comment]
On Thu, 1 Feb 2018, Adrian Schröter wrote:

> On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> > As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> > 
> > <constraints>
> > ...
> >   <overwrite>
> >    <conditions>
> >     <package>gcc8</package>
> >    </conditions>
> >    <hardware>
> >      <processors>4</processors>
> >      <physicalmemory>
> >       <size unit="G">2</size>
> >      </physicalmemory>
> >    </hardware>
> >   </overwrite>
> > </constraints>
> > 
> > thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> > memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> > get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> > 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> > the make parallelism that is used.
> > 
> > So please change the beahvior of pyhsicalmemory to scale with the number of threads
> > in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> > this behaivor.
> >
> > Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> > unreasonably (I don't remember if we can specify constraints per architecture).
> 
> It is the task of the spec file to specify a maximum number of parallel 
> jobs, but this is independend of the available hardware.

The .spec file just uses %{?_smp_mflags}

For GCC we want to use as many jobs as possible obviously since it's
a long time job in the base bootstrap cycle.  We could add
processors 16 and hard-code -j16 and require 8GB of memory but not sure
if that would be appreciated?

It should _really_ be the job of the scheduler to pick a worker
which matches the builds requirements but currently I simply cannot
specify the requirements accurately.

> Also I don't want to do an incompatible change here.

Then add a new constraint type.

**Comment by adrianschroeter** [comment]
On Donnerstag, 1. Februar 2018, 10:35:28 CET wrote rguenth:
> On Thu, 1 Feb 2018, Adrian Schröter wrote:
> 
> > On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> > > As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> > > 
> > > <constraints>
> > > ...
> > >   <overwrite>
> > >    <conditions>
> > >     <package>gcc8</package>
> > >    </conditions>
> > >    <hardware>
> > >      <processors>4</processors>
> > >      <physicalmemory>
> > >       <size unit="G">2</size>
> > >      </physicalmemory>
> > >    </hardware>
> > >   </overwrite>
> > > </constraints>
> > > 
> > > thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> > > memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> > > get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> > > 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> > > the make parallelism that is used.
> > > 
> > > So please change the beahvior of pyhsicalmemory to scale with the number of threads
> > > in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> > > this behaivor.
> > >
> > > Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> > > unreasonably (I don't remember if we can specify constraints per architecture).
> > 
> > It is the task of the spec file to specify a maximum number of parallel 
> > jobs, but this is independend of the available hardware.
> 
> The .spec file just uses %{?_smp_mflags}
> 
> For GCC we want to use as many jobs as possible obviously since it's
> a long time job in the base bootstrap cycle.  We could add
> processors 16 and hard-code -j16 and require 8GB of memory but not sure
> if that would be appreciated?

one way to go, but you can also calculate it yourself based on the provided 
memory. Maybe we could provide some rpm macro which helps here...

> It should _really_ be the job of the scheduler to pick a worker
> which matches the builds requirements but currently I simply cannot
> specify the requirements accurately.

The dispatcher has no chance, there are packages which only have a single
large file to compile (eg C++ gSOAP output) and the rest are smaller jobs.
We can not judge from the outside and the calculation would go wrong in this case.

> > Also I don't want to do an incompatible change here.
> 
> Then add a new constraint type.

we could do that, but I am no fan either of making it even more complicate...
However, will discuss it with my fellows.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by rguenth** [comment]
On Thu, 1 Feb 2018, Adrian Schröter wrote:

> On Donnerstag, 1. Februar 2018, 10:35:28 CET wrote rguenth:
> > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > 
> > > On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> > > > As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> > > > 
> > > > <constraints>
> > > > ...
> > > >   <overwrite>
> > > >    <conditions>
> > > >     <package>gcc8</package>
> > > >    </conditions>
> > > >    <hardware>
> > > >      <processors>4</processors>
> > > >      <physicalmemory>
> > > >       <size unit="G">2</size>
> > > >      </physicalmemory>
> > > >    </hardware>
> > > >   </overwrite>
> > > > </constraints>
> > > > 
> > > > thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> > > > memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> > > > get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> > > > 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> > > > the make parallelism that is used.
> > > > 
> > > > So please change the beahvior of pyhsicalmemory to scale with the number of threads
> > > > in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> > > > this behaivor.
> > > >
> > > > Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> > > > unreasonably (I don't remember if we can specify constraints per architecture).
> > > 
> > > It is the task of the spec file to specify a maximum number of parallel 
> > > jobs, but this is independend of the available hardware.
> > 
> > The .spec file just uses %{?_smp_mflags}
> > 
> > For GCC we want to use as many jobs as possible obviously since it's
> > a long time job in the base bootstrap cycle.  We could add
> > processors 16 and hard-code -j16 and require 8GB of memory but not sure
> > if that would be appreciated?
> 
> one way to go, but you can also calculate it yourself based on the provided 
> memory. Maybe we could provide some rpm macro which helps here...
> 
> > It should _really_ be the job of the scheduler to pick a worker
> > which matches the builds requirements but currently I simply cannot
> > specify the requirements accurately.
> 
> The dispatcher has no chance, there are packages which only have a single
> large file to compile (eg C++ gSOAP output) and the rest are smaller jobs.
> We can not judge from the outside and the calculation would go wrong in this case.
> 
> > > Also I don't want to do an incompatible change here.
> > 
> > Then add a new constraint type.
> 
> we could do that, but I am no fan either of making it even more complicate...
> However, will discuss it with my fellows.

The situation with the GCC build is the biggest memory consumer is
the link step of the compiler executables which consumes ~500MB
per executable.  The links happen all in parallel and we currently
build 8 such executables.  Maybe now all architectures we build
for have workers with at least 4GB of memory so we could raise
physicalmemory.  But I'm not sure for the arm32bit workers (actually
I'm quite sure those don't have 4GB).  In the docs I can't
see how I can specify architecture specific constraints (but
package specific constraints are also not documented...)

**Comment by adrianschroeter** [comment]
On Donnerstag, 1. Februar 2018, 10:49:22 CET wrote rguenth:
> On Thu, 1 Feb 2018, Adrian Schröter wrote:
> 
> > On Donnerstag, 1. Februar 2018, 10:35:28 CET wrote rguenth:
> > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > 
> > > > On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> > > > > As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> > > > > 
> > > > > <constraints>
> > > > > ...
> > > > >   <overwrite>
> > > > >    <conditions>
> > > > >     <package>gcc8</package>
> > > > >    </conditions>
> > > > >    <hardware>
> > > > >      <processors>4</processors>
> > > > >      <physicalmemory>
> > > > >       <size unit="G">2</size>
> > > > >      </physicalmemory>
> > > > >    </hardware>
> > > > >   </overwrite>
> > > > > </constraints>
> > > > > 
> > > > > thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> > > > > memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> > > > > get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> > > > > 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> > > > > the make parallelism that is used.
> > > > > 
> > > > > So please change the beahvior of pyhsicalmemory to scale with the number of threads
> > > > > in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> > > > > this behaivor.
> > > > >
> > > > > Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> > > > > unreasonably (I don't remember if we can specify constraints per architecture).
> > > > 
> > > > It is the task of the spec file to specify a maximum number of parallel 
> > > > jobs, but this is independend of the available hardware.
> > > 
> > > The .spec file just uses %{?_smp_mflags}
> > > 
> > > For GCC we want to use as many jobs as possible obviously since it's
> > > a long time job in the base bootstrap cycle.  We could add
> > > processors 16 and hard-code -j16 and require 8GB of memory but not sure
> > > if that would be appreciated?
> > 
> > one way to go, but you can also calculate it yourself based on the provided 
> > memory. Maybe we could provide some rpm macro which helps here...
> > 
> > > It should _really_ be the job of the scheduler to pick a worker
> > > which matches the builds requirements but currently I simply cannot
> > > specify the requirements accurately.
> > 
> > The dispatcher has no chance, there are packages which only have a single
> > large file to compile (eg C++ gSOAP output) and the rest are smaller jobs.
> > We can not judge from the outside and the calculation would go wrong in this case.
> > 
> > > > Also I don't want to do an incompatible change here.
> > > 
> > > Then add a new constraint type.
> > 
> > we could do that, but I am no fan either of making it even more complicate...
> > However, will discuss it with my fellows.
> 
> The situation with the GCC build is the biggest memory consumer is
> the link step of the compiler executables which consumes ~500MB
> per executable.  The links happen all in parallel and we currently
> build 8 such executables.  Maybe now all architectures we build
> for have workers with at least 4GB of memory so we could raise
> physicalmemory.  But I'm not sure for the arm32bit workers (actually
> I'm quite sure those don't have 4GB).

check the _statistics file of the build result how much got used ...

> In the docs I can't
> see how I can specify architecture specific constraints (but
> package specific constraints are also not documented...)

Using the overwrite element ... check 

 osc less openSUSE:Factory MozillaFirefox _constraints

for an example.


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by rguenth** [comment]
On Thu, 1 Feb 2018, Adrian Schröter wrote:

> On Donnerstag, 1. Februar 2018, 10:49:22 CET wrote rguenth:
> > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > 
> > > On Donnerstag, 1. Februar 2018, 10:35:28 CET wrote rguenth:
> > > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > > 
> > > > > On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> > > > > > As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> > > > > > 
> > > > > > <constraints>
> > > > > > ...
> > > > > >   <overwrite>
> > > > > >    <conditions>
> > > > > >     <package>gcc8</package>
> > > > > >    </conditions>
> > > > > >    <hardware>
> > > > > >      <processors>4</processors>
> > > > > >      <physicalmemory>
> > > > > >       <size unit="G">2</size>
> > > > > >      </physicalmemory>
> > > > > >    </hardware>
> > > > > >   </overwrite>
> > > > > > </constraints>
> > > > > > 
> > > > > > thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> > > > > > memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> > > > > > get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> > > > > > 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> > > > > > the make parallelism that is used.
> > > > > > 
> > > > > > So please change the beahvior of pyhsicalmemory to scale with the number of threads
> > > > > > in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> > > > > > this behaivor.
> > > > > >
> > > > > > Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> > > > > > unreasonably (I don't remember if we can specify constraints per architecture).
> > > > > 
> > > > > It is the task of the spec file to specify a maximum number of parallel 
> > > > > jobs, but this is independend of the available hardware.
> > > > 
> > > > The .spec file just uses %{?_smp_mflags}
> > > > 
> > > > For GCC we want to use as many jobs as possible obviously since it's
> > > > a long time job in the base bootstrap cycle.  We could add
> > > > processors 16 and hard-code -j16 and require 8GB of memory but not sure
> > > > if that would be appreciated?
> > > 
> > > one way to go, but you can also calculate it yourself based on the provided 
> > > memory. Maybe we could provide some rpm macro which helps here...
> > > 
> > > > It should _really_ be the job of the scheduler to pick a worker
> > > > which matches the builds requirements but currently I simply cannot
> > > > specify the requirements accurately.
> > > 
> > > The dispatcher has no chance, there are packages which only have a single
> > > large file to compile (eg C++ gSOAP output) and the rest are smaller jobs.
> > > We can not judge from the outside and the calculation would go wrong in this case.
> > > 
> > > > > Also I don't want to do an incompatible change here.
> > > > 
> > > > Then add a new constraint type.
> > > 
> > > we could do that, but I am no fan either of making it even more complicate...
> > > However, will discuss it with my fellows.
> > 
> > The situation with the GCC build is the biggest memory consumer is
> > the link step of the compiler executables which consumes ~500MB
> > per executable.  The links happen all in parallel and we currently
> > build 8 such executables.  Maybe now all architectures we build
> > for have workers with at least 4GB of memory so we could raise
> > physicalmemory.  But I'm not sure for the arm32bit workers (actually
> > I'm quite sure those don't have 4GB).
> 
> check the _statistics file of the build result how much got used ...

Ah, interesting.  I have to reverse engineer the worker configuration
though -- any chance to put information about the worker # cpus,
memory and disk size in that file?  The build log tells me only
the memory and cpu configuration via the qemu invocation command:

[   48s] /usr/bin/qemu-system-aarch64 -nodefaults -no-reboot -nographic 
-vga none -enable-kvm -cpu host -M virt,gic-version=host -object 
rng-random,filename=/dev/hwrng,id=rng0 -device virtio-rng-device,rng=rng0 
-runas qemu -net none -kernel 
/var/cache/obs/worker/root_2/.mount/boot/kernel -initrd 
/var/cache/obs/worker/root_2/.mount/boot/initrd -append 
root=/dev/disk/by-id/virtio-0 rootfstype=ext3 
rootflags=data=writeback,nobarrier,commit=150,noatime panic=1 quiet 
no-kvmclock nmi_watchdog=0 rw rd.driver.pre=binfmt_misc elevator=noop 
console=ttyAMA0 init=/.build/build -m 8192 -drive 
file=/var/cache/obs/worker/root_2/root,format=raw,if=none,id=disk,serial=0,cache=unsafe 
-device virtio-blk-device,drive=disk -drive 
file=/var/cache/obs/worker/root_2/swap,format=raw,if=none,id=swap,serial=1,cache=unsafe 
-device virtio-blk-device,drive=swap -serial stdio -smp 4


> > In the docs I can't
> > see how I can specify architecture specific constraints (but
> > package specific constraints are also not documented...)
> 
> Using the overwrite element ... check 
> 
>  osc less openSUSE:Factory MozillaFirefox _constraints
> 
> for an example.

Thanks.

**Comment by adrianschroeter** [comment]
On Donnerstag, 1. Februar 2018, 11:04:48 CET wrote rguenth:
> On Thu, 1 Feb 2018, Adrian Schröter wrote:
> 
> > On Donnerstag, 1. Februar 2018, 10:49:22 CET wrote rguenth:
> > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > 
> > > > On Donnerstag, 1. Februar 2018, 10:35:28 CET wrote rguenth:
> > > > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > > > 
> > > > > > On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> > > > > > > As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> > > > > > > 
> > > > > > > <constraints>
> > > > > > > ...
> > > > > > >   <overwrite>
> > > > > > >    <conditions>
> > > > > > >     <package>gcc8</package>
> > > > > > >    </conditions>
> > > > > > >    <hardware>
> > > > > > >      <processors>4</processors>
> > > > > > >      <physicalmemory>
> > > > > > >       <size unit="G">2</size>
> > > > > > >      </physicalmemory>
> > > > > > >    </hardware>
> > > > > > >   </overwrite>
> > > > > > > </constraints>
> > > > > > > 
> > > > > > > thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> > > > > > > memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> > > > > > > get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> > > > > > > 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> > > > > > > the make parallelism that is used.
> > > > > > > 
> > > > > > > So please change the beahvior of pyhsicalmemory to scale with the number of threads
> > > > > > > in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> > > > > > > this behaivor.
> > > > > > >
> > > > > > > Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> > > > > > > unreasonably (I don't remember if we can specify constraints per architecture).
> > > > > > 
> > > > > > It is the task of the spec file to specify a maximum number of parallel 
> > > > > > jobs, but this is independend of the available hardware.
> > > > > 
> > > > > The .spec file just uses %{?_smp_mflags}
> > > > > 
> > > > > For GCC we want to use as many jobs as possible obviously since it's
> > > > > a long time job in the base bootstrap cycle.  We could add
> > > > > processors 16 and hard-code -j16 and require 8GB of memory but not sure
> > > > > if that would be appreciated?
> > > > 
> > > > one way to go, but you can also calculate it yourself based on the provided 
> > > > memory. Maybe we could provide some rpm macro which helps here...
> > > > 
> > > > > It should _really_ be the job of the scheduler to pick a worker
> > > > > which matches the builds requirements but currently I simply cannot
> > > > > specify the requirements accurately.
> > > > 
> > > > The dispatcher has no chance, there are packages which only have a single
> > > > large file to compile (eg C++ gSOAP output) and the rest are smaller jobs.
> > > > We can not judge from the outside and the calculation would go wrong in this case.
> > > > 
> > > > > > Also I don't want to do an incompatible change here.
> > > > > 
> > > > > Then add a new constraint type.
> > > > 
> > > > we could do that, but I am no fan either of making it even more complicate...
> > > > However, will discuss it with my fellows.
> > > 
> > > The situation with the GCC build is the biggest memory consumer is
> > > the link step of the compiler executables which consumes ~500MB
> > > per executable.  The links happen all in parallel and we currently
> > > build 8 such executables.  Maybe now all architectures we build
> > > for have workers with at least 4GB of memory so we could raise
> > > physicalmemory.  But I'm not sure for the arm32bit workers (actually
> > > I'm quite sure those don't have 4GB).
> > 
> > check the _statistics file of the build result how much got used ...
> 
> Ah, interesting.  I have to reverse engineer the worker configuration
> though -- any chance to put information about the worker # cpus,
> memory and disk size in that file?  The build log tells me only
> the memory and cpu configuration via the qemu invocation command:

not sure if it is needed to put it in that file, you can request it via

 # osc workerinfo x86_64:lamb01:1

for example.


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by rguenth** [comment]
On Thu, 1 Feb 2018, Adrian Schröter wrote:

> On Donnerstag, 1. Februar 2018, 11:04:48 CET wrote rguenth:
> > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > 
> > > On Donnerstag, 1. Februar 2018, 10:49:22 CET wrote rguenth:
> > > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > > 
> > > > > On Donnerstag, 1. Februar 2018, 10:35:28 CET wrote rguenth:
> > > > > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > > > > 
> > > > > > > On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> > > > > > > > As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> > > > > > > > 
> > > > > > > > <constraints>
> > > > > > > > ...
> > > > > > > >   <overwrite>
> > > > > > > >    <conditions>
> > > > > > > >     <package>gcc8</package>
> > > > > > > >    </conditions>
> > > > > > > >    <hardware>
> > > > > > > >      <processors>4</processors>
> > > > > > > >      <physicalmemory>
> > > > > > > >       <size unit="G">2</size>
> > > > > > > >      </physicalmemory>
> > > > > > > >    </hardware>
> > > > > > > >   </overwrite>
> > > > > > > > </constraints>
> > > > > > > > 
> > > > > > > > thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> > > > > > > > memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> > > > > > > > get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> > > > > > > > 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> > > > > > > > the make parallelism that is used.
> > > > > > > > 
> > > > > > > > So please change the beahvior of pyhsicalmemory to scale with the number of threads
> > > > > > > > in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> > > > > > > > this behaivor.
> > > > > > > >
> > > > > > > > Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> > > > > > > > unreasonably (I don't remember if we can specify constraints per architecture).
> > > > > > > 
> > > > > > > It is the task of the spec file to specify a maximum number of parallel 
> > > > > > > jobs, but this is independend of the available hardware.
> > > > > > 
> > > > > > The .spec file just uses %{?_smp_mflags}
> > > > > > 
> > > > > > For GCC we want to use as many jobs as possible obviously since it's
> > > > > > a long time job in the base bootstrap cycle.  We could add
> > > > > > processors 16 and hard-code -j16 and require 8GB of memory but not sure
> > > > > > if that would be appreciated?
> > > > > 
> > > > > one way to go, but you can also calculate it yourself based on the provided 
> > > > > memory. Maybe we could provide some rpm macro which helps here...
> > > > > 
> > > > > > It should _really_ be the job of the scheduler to pick a worker
> > > > > > which matches the builds requirements but currently I simply cannot
> > > > > > specify the requirements accurately.
> > > > > 
> > > > > The dispatcher has no chance, there are packages which only have a single
> > > > > large file to compile (eg C++ gSOAP output) and the rest are smaller jobs.
> > > > > We can not judge from the outside and the calculation would go wrong in this case.
> > > > > 
> > > > > > > Also I don't want to do an incompatible change here.
> > > > > > 
> > > > > > Then add a new constraint type.
> > > > > 
> > > > > we could do that, but I am no fan either of making it even more complicate...
> > > > > However, will discuss it with my fellows.
> > > > 
> > > > The situation with the GCC build is the biggest memory consumer is
> > > > the link step of the compiler executables which consumes ~500MB
> > > > per executable.  The links happen all in parallel and we currently
> > > > build 8 such executables.  Maybe now all architectures we build
> > > > for have workers with at least 4GB of memory so we could raise
> > > > physicalmemory.  But I'm not sure for the arm32bit workers (actually
> > > > I'm quite sure those don't have 4GB).
> > > 
> > > check the _statistics file of the build result how much got used ...
> > 
> > Ah, interesting.  I have to reverse engineer the worker configuration
> > though -- any chance to put information about the worker # cpus,
> > memory and disk size in that file?  The build log tells me only
> > the memory and cpu configuration via the qemu invocation command:
> 
> not sure if it is needed to put it in that file, you can request it via
> 
>  # osc workerinfo x86_64:lamb01:1
> 
> for example.

Sure.  But the name of the worker isn't in the file either (and only
in the log?) ;)

**Comment by adrianschroeter** [comment]
On Donnerstag, 1. Februar 2018, 11:24:20 CET wrote rguenth:
> On Thu, 1 Feb 2018, Adrian Schröter wrote:
> 
> > On Donnerstag, 1. Februar 2018, 11:04:48 CET wrote rguenth:
> > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > 
> > > > On Donnerstag, 1. Februar 2018, 10:49:22 CET wrote rguenth:
> > > > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > > > 
> > > > > > On Donnerstag, 1. Februar 2018, 10:35:28 CET wrote rguenth:
> > > > > > > On Thu, 1 Feb 2018, Adrian Schröter wrote:
> > > > > > > 
> > > > > > > > On Donnerstag, 1. Februar 2018, 10:04:50 CET wrote rguenth:
> > > > > > > > > As working currently physicalmemory is quite useless to tell the system build time memory requirements given there's no control over how many CPUs the worker will get.  For example for gcc we have
> > > > > > > > > 
> > > > > > > > > <constraints>
> > > > > > > > > ...
> > > > > > > > >   <overwrite>
> > > > > > > > >    <conditions>
> > > > > > > > >     <package>gcc8</package>
> > > > > > > > >    </conditions>
> > > > > > > > >    <hardware>
> > > > > > > > >      <processors>4</processors>
> > > > > > > > >      <physicalmemory>
> > > > > > > > >       <size unit="G">2</size>
> > > > > > > > >      </physicalmemory>
> > > > > > > > >    </hardware>
> > > > > > > > >   </overwrite>
> > > > > > > > > </constraints>
> > > > > > > > > 
> > > > > > > > > thus guarantees a minimum amount of 4 CPUs and a minimum amount of 2GB of physical
> > > > > > > > > memory.  But in case we get a worker with 8 CPUs and 16 threads allocated we don't
> > > > > > > > > get the minimum amount of memory bumped to (2/4)*8.  It's also not clear whether
> > > > > > > > > 'processors' includes threads and thus matches %{jobs} will evaluate to and thus
> > > > > > > > > the make parallelism that is used.
> > > > > > > > > 
> > > > > > > > > So please change the beahvior of pyhsicalmemory to scale with the number of threads
> > > > > > > > > in the VM and or add a new pyhsicalmemoryperthread _constraint specifier matching
> > > > > > > > > this behaivor.
> > > > > > > > >
> > > > > > > > > Otherwise we have to "guess" the largest #CPU VMs and bump physicalmemory up
> > > > > > > > > unreasonably (I don't remember if we can specify constraints per architecture).
> > > > > > > > 
> > > > > > > > It is the task of the spec file to specify a maximum number of parallel 
> > > > > > > > jobs, but this is independend of the available hardware.
> > > > > > > 
> > > > > > > The .spec file just uses %{?_smp_mflags}
> > > > > > > 
> > > > > > > For GCC we want to use as many jobs as possible obviously since it's
> > > > > > > a long time job in the base bootstrap cycle.  We could add
> > > > > > > processors 16 and hard-code -j16 and require 8GB of memory but not sure
> > > > > > > if that would be appreciated?
> > > > > > 
> > > > > > one way to go, but you can also calculate it yourself based on the provided 
> > > > > > memory. Maybe we could provide some rpm macro which helps here...
> > > > > > 
> > > > > > > It should _really_ be the job of the scheduler to pick a worker
> > > > > > > which matches the builds requirements but currently I simply cannot
> > > > > > > specify the requirements accurately.
> > > > > > 
> > > > > > The dispatcher has no chance, there are packages which only have a single
> > > > > > large file to compile (eg C++ gSOAP output) and the rest are smaller jobs.
> > > > > > We can not judge from the outside and the calculation would go wrong in this case.
> > > > > > 
> > > > > > > > Also I don't want to do an incompatible change here.
> > > > > > > 
> > > > > > > Then add a new constraint type.
> > > > > > 
> > > > > > we could do that, but I am no fan either of making it even more complicate...
> > > > > > However, will discuss it with my fellows.
> > > > > 
> > > > > The situation with the GCC build is the biggest memory consumer is
> > > > > the link step of the compiler executables which consumes ~500MB
> > > > > per executable.  The links happen all in parallel and we currently
> > > > > build 8 such executables.  Maybe now all architectures we build
> > > > > for have workers with at least 4GB of memory so we could raise
> > > > > physicalmemory.  But I'm not sure for the arm32bit workers (actually
> > > > > I'm quite sure those don't have 4GB).
> > > > 
> > > > check the _statistics file of the build result how much got used ...
> > > 
> > > Ah, interesting.  I have to reverse engineer the worker configuration
> > > though -- any chance to put information about the worker # cpus,
> > > memory and disk size in that file?  The build log tells me only
> > > the memory and cpu configuration via the qemu invocation command:
> > 
> > not sure if it is needed to put it in that file, you can request it via
> > 
> >  # osc workerinfo x86_64:lamb01:1
> > 
> > for example.
> 
> Sure.  But the name of the worker isn't in the file either (and only
> in the log?) ;)

We could extend the osc command to find the last succeeded or failed worker and show it ...

Something for the osc github issues.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by scarabeusiv** [comment]
I guess the macro expansion could take into account something like MEM_PER_THREAD in constraints and then we could limit the amount of smp_mflags determined to be there:

Currently it is calculated as follows:
-13: _smp_mflags        %([ -z "$RPM_BUILD_NCPUS" ] \
        && RPM_BUILD_NCPUS="`/usr/bin/getconf _NPROCESSORS_ONLN`"; \
        ncpus_max=%{?_smp_ncpus_max}; \
        if [ -n "$ncpus_max" ] && [ "$ncpus_max" -gt 0 ] && [ "$RPM_BUILD_NCPUS" -gt "$ncpus_max" ]; then RPM_BUILD_NCPUS="$ncpus_max"; fi; \
        if [ "$RPM_BUILD_NCPUS" -gt 1 ]; then echo "-j$RPM_BUILD_NCPUS"; fi)

I guess it could be determined by build script to reduce the amount of jobs...

**Comment by rguenth** [comment]
Duplicate of #1953
